buildengine/build_kotlin.go

```diff
@@ -72,3 +108,180 @@ func SetPOMProperties(ctx context.Context, baseDir string) error {
 
 	return tree.WriteToFile(pomFile)
 }
+
+func prepareFTLRoot(module Module) error {
+	buildDir := filepath.Join(module.Dir, "target")
+	if err := os.MkdirAll(buildDir, 0700); err != nil {
+		return err
+	}
+
+	fileContent := fmt.Sprintf(`
+SchemaExtractorRuleSet:
+  ExtractSchemaRule:
+    active: true
+    output: %s
+`, buildDir)
+
+	detektYmlPath := filepath.Join(buildDir, "detekt.yml")
+	if err := os.WriteFile(detektYmlPath, []byte(fileContent), 0600); err != nil {
+		return fmt.Errorf("unable to configure detekt for %s: %w", module.Module, err)
+	}
+
+	mainFilePath := filepath.Join(buildDir, "main")
+
+	mainFile := `#!/bin/bash
+exec java -cp "classes:$(cat classpath.txt)" xyz.block.ftl.main.MainKt
+`
+	if err := os.WriteFile(mainFilePath, []byte(mainFile), 0700); err != nil { //nolint:gosec
+		return fmt.Errorf("unable to configure main executable for %s: %w", module.Module, err)
+	}
+	return nil
+}
+
+func generateExternalModules(ctx context.Context, module Module, sch *schema.Schema) error {
+	logger := log.FromContext(ctx)
+	config := module.ModuleConfig
+	funcs := maps.Clone(scaffoldFuncs)
+
+	// Wipe the modules directory to ensure we don't have any stale modules.
+	_ = os.RemoveAll(filepath.Join(config.Dir, "target", "generated-sources", "ftl"))
+
+	logger.Debugf("Generating external modules")
+	return internal.ScaffoldZip(kotlinruntime.ExternalModuleTemplates(), config.Dir, externalModuleContext{
+		module: module,
+		Schema: sch,
+	}, scaffolder.Functions(funcs))
+}
+
+var scaffoldFuncs = scaffolder.FuncMap{
```

Comment:
These functions are already defined in a couple of places. Let's factor them into a single location and include them by default in `internal.ScaffoldZip`.

Reply:
`scaffolder.Functions()` is cumulative, so the bespoke functions can still be added in each individual location.